### ISSUE TITLE
Recall reports a different class set than Precision and F1Score for the same data

### ISSUE THREAD

**Issue by arthi-arumugam-git** [roboflow/supervision#2467]
`Precision`, `F1Score` and `Recall` return `matched_classes` and a `*_per_class` array that
look like parallel outputs of the same computation. On identical input they are not, and
nothing raises.

## Reproduction

Ground truth contains only class 0. The model predicts class 0 correctly and also predicts
class 1, which never appears in the ground truth at all.

```python
import numpy as np
import supervision as sv
from supervision.metrics import F1Score, Precision, Recall, MetricTarget, AveragingMethod

targets = sv.Detections(
    xyxy=np.array([[10., 10., 50., 50.]]),
    class_id=np.array([0]),
)
predictions = sv.Detections(
    xyxy=np.array([[10., 10., 50., 50.], [100., 100., 140., 140.]]),
    class_id=np.array([0, 1]),
    confidence=np.array([0.9, 0.9]),
)

for name, cls in (("precision", Precision), ("recall", Recall), ("f1", F1Score)):
    m = cls(metric_target=MetricTarget.BOXES, averaging_method=AveragingMethod.MACRO)
    m.update(predictions, targets)
    r = m.compute()
    print(name, r.matched_classes)
```

Observed on `develop` (supervision 0.29.1):

```
precision  [0 1]      precision_per_class shape (2, 10)
recall     [0]        recall_per_class    shape (1, 10)
f1         [0 1]
```

The practical consequence is that the obvious thing to do with these results is wrong:

```python
for cid, p, r in zip(pres.matched_classes, pres.precision_per_class[:, 0],
                     rec.recall_per_class[:, 0]):
    ...
```

`zip` truncates to one row while precision reported two classes, and the hallucinated class
disappears from the table with no error. If the class ordering had differed rather than the
length, the values would have been attributed to the wrong classes instead.

## Where it comes from

#2331, "count false positives on background images and absent classes", changed
`f1_score.py` and `precision.py` and added a regression test to each. It did not touch
`recall.py`. The line it replaced in `precision.py`:

```python
unique_classes, class_counts = np.unique(true_class_ids, return_counts=True)
```

is still present at `src/supervision/metrics/recall.py:331`.

`tests/metrics/test_recall.py` is also the only metric test file with no coverage for a
class that appears solely in predictions. `test_detection.py`, `test_f1_score.py`,
`test_mean_average_recall.py` and `test_precision.py` all have it.

The two docstrings have diverged accordingly. `PrecisionResult.matched_classes` says "the
class IDs present in either predictions or ground truth ... Classes that appear only in
predictions (no ground-truth instances) are included". `RecallResult.matched_classes` says
"the class IDs of all matched classes", with `recall_per_class` documented as
`(num_target_classes, num_iou_thresholds)`.

## The question, because I do not think this is obvious

Recall for a class with no ground-truth instances is undefined, not zero. Including such a
class would add a 0.0 into the macro mean and lower reported recall, which is defensible for
precision (those predictions are real false positives) and arguably wrong for recall.

So there are two reasonable fixes and they are not equivalent:

1. **Match `Precision`.** `unique_classes` becomes the union, and macro recall changes for
   anyone whose model predicts classes absent from the ground truth. Consistent, but it
   moves a published number.
2. **Keep recall's scoring and fix only the contract.** Leave the macro value alone and make
   the difference explicit, so the per-class arrays are never silently mismatched.

Happy to send the PR with tests for whichever you prefer. I would lean to (2) plus an
explicit note in `RecallResult`, but this is your metric to define and I did not want to
change a number you publish on a guess.

**Comment by arthi-arumugam-git** [comment]
Answering my own question above, because two things in the repo settle it and I would rather
show the reasoning than just push a diff.

`test_precision.py` already cites sklearn consistency as the standard for the `WEIGHTED`
case. sklearn infers labels from the union of `y_true` and `y_pred`, and for a label present
only in predictions it reports per-class recall `0.0` and includes it in the macro average.
So option 1 is the one that matches both `Precision` and the standard this repo already
chose.

The second thing is more direct: `recall.py` already carries #2331's `WEIGHTED` zero-support
guard, and its comment refers to "only false-positive classes". That state cannot occur in
`recall.py` as written, because `unique_classes` came from the ground truth alone. The guard
was propagated across; the union that gives it meaning was not.

PR is #2468. `MICRO` and `WEIGHTED` are unchanged and `MACRO` moves, which the changelog
entry states rather than buries. Whole metrics suite 280 passed, up from 276, with no
existing test modified.

**Comment by Daniiiil1** [comment]
Hi — I can turn the analysis in this issue into a focused PR: use the union of `y_true` and `y_pred` as the shared class universe, keep Precision/Recall/F1 ordering consistent, and add sklearn-based golden tests for labels present only in ground truth, only in predictions, empty cases, and macro/weighted averages. If nobody is already implementing it, would this direction be accepted?

— Daniil Mordanov

**Comment by arthi-arumugam-git** [comment]
Hi Daniil, thanks for offering, and your direction is right. It is the same conclusion I reached above, for the same reason: sklearn infers labels from the union, so a prediction-only label gets recall `0.0` and counts in the macro average.

It is already implemented in #2468, which is open and under review. `Recall._compute_recall_for_classes` builds the class set from the union, background images (predictions with no targets) are handled the way `Precision` does after a review catch from Copilot, `RecallResult` is documented to match `PrecisionResult`, and the changelog states the `MACRO` change rather than burying it. `MICRO` and `WEIGHTED` are unchanged. Suite is 281 passing, up from 276, with no existing test modified.

The part of your proposal that #2468 does **not** cover is worth doing separately, and I would rather you took it than duplicated this: **golden tests against sklearn itself.** Right now the repo cites sklearn as the standard in a comment in `test_precision.py`, but nothing actually asserts parity. A test module that runs `sv.Precision`, `sv.Recall` and `sv.F1Score` against `sklearn.metrics` over a matrix of cases (labels only in ground truth, only in predictions, both, empty, and each averaging method) would turn that citation into something enforced. That is a real gap and it does not touch the same files.

Happy to review it if you open one.

**Comment by Daniiiil1** [comment]
Thanks — I’ll take the separate sklearn parity test module. I’ll cover prediction-only labels, target-only labels, background images, empty inputs, non-contiguous class IDs, and MICRO/MACRO/WEIGHTED averaging for Precision, Recall, and F1.

Since Recall parity depends on #2468, I’ll develop against that behavior and rebase onto `develop` after it is merged. Is adding `scikit-learn` to the development/test dependency group acceptable, or would you prefer a different CI arrangement?

— Daniil Mordanov

**Comment by arthi-arumugam-git** [comment]
That scope sounds right, and non-contiguous class IDs is a good addition I had not thought of.

To be clear though, I am not a maintainer here, so I cannot tell you whether a new test dependency is acceptable. That is a call for @SkalskiP or @Borda.

What I would say as an opinion: I would try to avoid the dependency. `supervision` deliberately keeps its install light, and even a dev-group addition puts scikit-learn plus scipy into every CI matrix cell, and this repo runs a wide one (three operating systems, Python 3.10 through 3.13, two opencv variants). That is a real cost for a test module.

The alternative that gets the same guarantee: compute sklearn's outputs once, locally, and commit them as golden constants with the generating snippet in a docstring beside them. The parity is then pinned in the repository rather than recomputed on every run, it is diffable when a value changes, and CI stays as light as it is now. The obvious objection is that the constants go stale if sklearn changes its behaviour, but that is exactly the kind of change you would want to see as a failing test and a visible diff rather than silently absorbed.

Either way it is their decision, not mine.

On the ordering: #2468 is open and its checks are green, so developing against that behaviour makes sense. Just be aware that the `MACRO` recall value for a prediction-only class moves under it, from being excluded to contributing `0.0`. `MICRO` and `WEIGHTED` do not move. If you pin golden values before it merges, that is the one number that will need regenerating afterwards.
